lib/services/ngsi/entities-NGSI-LD.js

```diff
@@ -0,0 +1,541 @@
+            }
+        } else {
+            logger.debug(context, 'Unknown error executing ' + operationName + ' operation');
+            if (!(body instanceof Array || body instanceof Object)) {
+                body = JSON.parse(body);
+            }
+
+            callback(new errors.EntityGenericError(entityName, typeInformation.type, body, response.statusCode));
+        }
+    };
+}
+
+/**
+ * Makes a query to the Device's entity in the context broker using NGSI-LD, with the list
+ * of attributes given by the 'attributes' array.
+ *
+ * @param {String} entityName       Name of the entity to query.
+ * @param {Array} attributes        Attribute array containing the names of the attributes to query.
+ * @param {Object} typeInformation  Configuration information for the device.
+ * @param {String} token            User token to identify against the PEP Proxies (optional).
+ */
+function sendQueryValueNgsiLD(entityName, attributes, typeInformation, token, callback) {
+    let url = '/ngsi-ld/v1/entities/urn:ngsi-ld:' + typeInformation.type + ':' + entityName;
+
+    if (attributes && attributes.length > 0) {
+        url = url + '?attrs=' + attributes.join(',');
+    }
+
+    const options = NGSIUtils.createRequestObject(url, typeInformation, token);
+    options.method = 'GET';
+    options.json = true;
+
+    if (!typeInformation || !typeInformation.type) {
+        callback(new errors.TypeNotFound(null, entityName));
+        return;
+    }
+
+    logger.debug(context, 'Querying values of the device in the Context Broker at [%s]', options.url);
+    logger.debug(context, 'Using the following request:\n\n%s\n\n', JSON.stringify(options, null, 4));
+
+    request(
+        options,
+        generateNGSILDOperationHandler('query', entityName, typeInformation, token, options, function(error, result) {
+            if (error) {
+                callback(error);
+            } else {
+                NGSIUtils.applyMiddlewares(NGSIUtils.queryMiddleware, result, typeInformation, callback);
+            }
+        })
+    );
+}
+
+/**
+ * Makes an update in the Device's entity in the context broker, with the values given in the 'attributes' array.
+ * This array should comply to the NGSI-LD's attribute format.
+ *
+ * @param {String} entityName       Name of the entity to register.
+ * @param {Array} attributes        Attribute array containing the values to update.
+ * @param {Object} typeInformation  Configuration information for the device.
+ * @param {String} token            User token to identify against the PEP Proxies (optional).
+ */
+function sendUpdateValueNgsiLD(entityName, attributes, typeInformation, token, callback) {
+    let payload = {};
+
+    /*var url = '/ngsi-ld/v1/entities/' + entityName + '/attrs';
+
+    if (typeInformation.type) {
+       url += '?type=' + typeInformation.type;
+    }*/
+
+    const url = '/ngsi-ld/v1/entityOperations/upsert/';
```

Comment:
Why did you choose to do an ```/entityOperations/upsert``` over a ```PATCH /ngsi-ld/v1/entities/<entityId>/attrs``` on the entity to be updated ?

Reply:
Just following the principle of changing the underlying code as little as possible. The *NSGI v2* Interface was achieved using batch commands so the *NGSI-LD* interface follows the same paradigm.